explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,260 @@
+```
+
+Note how `file` reports the binary as being `with debug_info, not stripped`.  This is the output you need to look for when inspecting your own binaries.
+
+You can use the `eu-readelf` command (from the `elfutils` package) to determine whether any compression was performed with `dwz`:
+
+```bash
+$ eu-readelf -S /usr/bin/openssl
+[...]
+Section Headers:
+[Nr] Name                 Type         Addr             Off      Size     ES Flags Lk Inf Al
+[...]
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` (as in `compressed`) in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build.
+
+### Using `perf` to obtain profile data
+
+With OpenSSL prepared and ready to be profiled, it's now time to use `perf` to obtain profiling data from our workload.  First, let's make sure we can actually access profiling data in the system.  As `root`, run:
+
+```bash
+# echo -1 > /proc/sys/kernel/perf_event_paranoid
+```
+
+This should allow all users to monitor events in the system.
+
+Then, if you are on an Intel system, you can invoke `perf` using:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.
+
+Note that the only thing that differs between the Intel and AMD variants is the PMU event to be monitored.  Also, note how we are using `sudo` to invoke `perf record`.  This is necessary in order to obtain full access to Linux kernel symbols and relocation information.  It also means that the `openssl-nonpgo.perfdata` file ownership will need to be adjusted:
+
+```bash
+$ sudo chown ubuntu:ubuntu openssl-nonpgo.perfdata
+```
+
+Now, we need to convert the file to `gcov`.
+
+#### Alternative scenario: profiling the program using other approaches
+
+We do not always have well behaved programs that can be profiled like `openssl speed`.  Sometimes, the program might take a long time to finish running the workload, which ends up generating huge `perfdata` files that can prove very challenging for `autofdo` to process.  For example, we profiled QEMU internally and, when executing a workload that took around 3 hours to complete, the `perfdata` generated had a size of several gigabytes.
```

Comment:
When you say "qemu internally" you might also refer to the related blog post.

Reply:
+1.  I will publish this documentation first, but I'll amend it later with links to the blog post.

Reply:
Done.